python/lsst/daf/butler/core/butlerConfig.py

```diff
@@ -51,54 +63,64 @@ def __init__(self, other):
         butlerConfig = Config(other)
 
-        # Check that fundamental keys are present
-        self._validate(butlerConfig)
+        # All the configs that can be associated with default files
+        configComponents = (SchemaConfig, StorageClassConfig, DatastoreConfig, RegistryConfig)
+
+        # This is a list of all the config files to search
+        defaultsFiles = [(c, c.defaultConfigFile) for c in configComponents]
 
-        # Look for class specific defaults
-        for section in ("datastore", "registry"):
-            k = f"{section}.cls"
-            print("Checking {}: {}".format(k, butlerConfig[k]))
+        # Components that derive their configurations from implementation
+        # classes rather than configuration classes
+        specializedConfigs = (DatastoreConfig, RegistryConfig)
+
+        # Look for class specific default files to be added to the file list
+        # These class definitions must come from the suppliedbutler config and
+        # are not defaulted.
+        for componentConfig in specializedConfigs:
+            k = "{}.cls".format(componentConfig.component)
+            if k not in butlerConfig:
+                raise ValueError("Required configuration {} not present in {}".format(k, other))
             cls = doImport(butlerConfig[k])
-            defaultsPath = cls.defaults
-            if defaultsPath is not None:
-                if not os.path.isabs(defaultsPath):
-                    defaultsPath = os.path.join(self.defaultsDir, defaultsPath)
-                c = Config(defaultsPath)
-                # Merge into baseline
-                self.update(c)
+            defaultsFile = cls.defaultConfigFile
+            if defaultsFile is not None:
+                defaultsFiles.append((componentConfig, defaultsFile))
+
+        # We can pick up defaults from multiple search paths
+        # We fill defaults by using the butler config path and then
+        # the config path environment variable in reverse order.
+        self.defaultsPaths = []
+
+        # Find the butler configs
+        if "DAF_BUTLER_DIR" in os.environ:
+            self.defaultsPaths.append(os.path.join(os.environ["DAF_BUTLER_DIR"], "config"))
+
+        if "DAF_BUTLER_CONFIG_PATH" in os.environ:
+            externalPaths = list(reversed(os.environ["DAF_BUTLER_CONFIG_PATH"].split(os.pathsep)))
+            self.defaultsPaths.append(externalPaths)
+
+        # Search each directory for each of the default files
+        for pathDir in self.defaultsPaths:
+            for configClass, configFile in defaultsFiles:
+                # Assume external paths have same config files as global config
+                # directory. Absolute paths are possible for external
+                # code.
+                # Should this be a log message? Are we using lsst.log?
+                # print("Checking path {} for {} ({})".format(pathDir, configClass, configFile))
+                if not os.path.isabs(configFile):
```

Comment:
Doesn't this apply absolute paths over and over?  Maybe that's not a problem.

Reply:
Yes. It does. Hadn't thought of that. I don't think it's a problem as such, but it will slow things down.

Reply:
Why not simply invert the loop order?

Reply:
yes. since the configs read from the paths are subsets of the config I can loop over configs and then paths.